ENH: use analytic formula for log-laplace MLE when loc is known.

fancidev opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message

Depending on your changes, you can skip CI operations and save time and energy: 
http://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
Closes gh-19996.

#### What does this implement/fix?
When `floc` is provided, use a closed-form formula to compute the MLE for log-laplace distribution. The estimator is consistent.

#### Additional information
MLE when `floc` is unknown is discussed in gh-19996, and is shown to be inconsistent in some cases. Therefore the MLE is not specialised for `floc` unknown.

Separately, closed-form expressions for the moment estimator when `floc` is known are given in gh-19996, and are shown to be inconsistent for many parameter combinations. Therefore MM estimator should not be used for log-laplace in practice.

fancidev committed:
ENH: use analytic formula for log-laplace MLE when loc is known.

mdhaber commented on scipy/stats/_continuous_distns.py:
```diff
@@ -6384,6 +6384,37 @@ def _munp(self, n, c):
+        # When location is zero, the log-Laplace distribution is related to
+        # the Laplace distribution in that if X ~ Laplace(loc=a, scale=b),
+        # then Y = exp(X) ~ LogLaplace(c=1/b, loc=0, scale=exp(a)).  It can
+        # be shown that the MLE for Y is the same as the MLE for X = ln(Y).
```
I think this is true in general for log transformed distributions due to invariance properties of MLE, right?

What do you think of calling `laplace.fit` instead of repeating the formulas?

If they were any more complicated I think it would be more important, but I suppose if we do we'd have to transform parameters back and forth. Up to you.

fancidev replied:
Yes it’s in general true for monotone transform of the random variable where the transform itself doesn’t involve unknown parameters.

It’s probably better to call `laplace.fit` instead, in case the latter gains more features in the future. It also helps to reduce the surface for testing.

mdhaber commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -3089,6 +3089,46 @@ def test_moments_stats(self, r):
+        # Sanity check that the specialized MLE is actually executed.
+        # Due to floating point arithmetic, the generic MLE is unlikely
+        # to produce the exact same result as the analytic MLE.
+        assert np.any(generic_args != analytic_args)
```
There is a function [`assert_less_or_close_loglike`](https://github.com/scipy/scipy/blob/4edfcaa3ce8a387450b6efce968572def71be089/scipy/stats/tests/test_distributions.py#L251) that takes care of the comparison. Could you add this check that they are not exactly equal to it (and use it)?

fancidev replied:
Sure let me do that.

fancidev replied:
Got 44 test failures after that change… I’ll look into whether they’re expected (some methods fall back to super mle when it finds difficulty) or the sanity check is too speculative.

mdhaber replied:
Hmm that would make sense. Could make that check opt-in.

fancidev replied:
Done!

mdhaber commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -3089,6 +3089,46 @@ def test_moments_stats(self, r):
+    @pytest.mark.parametrize("c", [0.5, 1.0, 2.0])
+    @pytest.mark.parametrize("loc, scale", [(-1.2, 3.45)])
+    @pytest.mark.parametrize("fix_c", [True, False])
+    @pytest.mark.parametrize("fix_loc", [True])
```
It sounds like we're not going to override the generic method, so maybe just set `floc` in the function?

fancidev replied:
Sure I will update that!

fancidev committed:
ENH: reuse laplace.fit() in loglaplace.fit() when floc is given.

fancidev committed:
TST: use utility function to check log likelihood of analytic mle.

fancidev committed:
TST: add sanity check to _assert_less_or_close_loglike().

The sanity check asserts that the analytical fit result is not identical
to the numerical fit result.  This is usually true due to floating point
arithmetic.

The check may be bypassed by passing maybe_identical=True.  This could
happen if the analytical routine falls back to numerical MLE in some
cases, or if the objective function is simple enough that the numerical
solution and the analytical solution coincide.

VascoSch92 commented on scipy/stats/_continuous_distns.py:
```diff
@@ -6384,6 +6384,38 @@ def _munp(self, n, c):
+        # be shown that the MLE for Y is the same as the MLE for X = ln(Y).
+        # Therefore, we reuse the formulas from laplace.fit() and transform
+        # the result back into log-laplace's parameter space.
+        a, b = laplace.fit(np.log(data),
```
```python
a, b = laplace.fit(
    np.log(data),
    floc=np.log(fscale) if fscale is not None else None,
    fscale=1/fc if fc is not None else None,
    method='mle',
)
```

fancidev replied:
Same as above.

VascoSch92 commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -261,6 +262,13 @@ def _assert_less_or_close_loglike(dist, data, func=None, **kwds):
+    # Sanity check that the analytical MLE is actually executed.
+    # Due to floating point arithmetic, the generic MLE is unlikely
+    # to produce the exact same result as the analytical MLE.
+    if not maybe_identical:
```
`if maybe_identical is True` easy to read and more udnerstandable

fancidev replied:
Do you mean `if maybe_identical is False`?

VascoSch92 replied:
Yes exactly sorry :-) But this is the point. `if maybe_identical is False` is much easier to understand

fancidev replied:
The general recommendation from [Python style guide](https://peps.python.org/pep-0008/) is “Don’t compare boolean values to True or False using == [or is]”. Though in this case I agree “not” followed by “maybe” sounds like double negation and not entirely straightforward. Since this is just in the test and not in the main code, I’d prefer to leave it as is to follow the style guide.

fancidev commented on scipy/stats/_continuous_distns.py:
```diff
@@ -6384,6 +6384,38 @@ def _munp(self, n, c):
 
+    @_call_super_mom
+    def fit(self, data, *args, **kwds):
+        data, fc, floc, fscale = _check_fit_input_parameters(self, data,
```
The code style you suggest is more modern and maybe popular. I just followed the code style of the existing scipy code base, which is more compact though less uniform looking.

lucascolley replied:
In cases like this either is fine - style is not uniform across the repo and we allow for author preference to a reasonable extent (i.e. we don't spend review time on it).

_Hopefully one day we can get an auto-formatter, but there are challenges with that and for now it's fine and it would not be a good idea to start nitpicking style in every PR_

fancidev commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -1177,7 +1185,10 @@ def test_fit_MLE_comp_optimizer(self, rvs_loc, rvs_scale,
-        _assert_less_or_close_loglike(stats.halfnorm, data, **kwds)
+        # Numerical result may equal analytical result if the initial guess
+        # computed from moment condition is already optimal.
+        _assert_less_or_close_loglike(stats.halfnorm, data, **kwds,
```
Here the keyword arguments are a known dict used to pass a subset of fixed parameters, and they conceptually “bind tighter” with the main functionality of the function. The additional keyword argument “maybe_identical” is kind of an “option”. I put it after the **kwds as it is common to put optional arguments in the end.

mdhaber commented on scipy/stats/_continuous_distns.py:
```diff
@@ -6384,6 +6384,38 @@ def _munp(self, n, c):
+                           floc=np.log(fscale) if fscale is not None else None,
+                           fscale=1/fc if fc is not None else None,
+                           method='mle')
+        if fscale is None:
+            fscale = np.exp(a)
+        if fc is None:
+            fc = 1 / b
+        return fc, floc, fscale
```
There has been some confusion in the past about `fscale` vs `scale`, etc., so I try to reserve `f`-prefixed variable names for values that are fixed.

```suggestion
        loc = floc
        scale = np.exp(a) if fscale is None else fscale
        c = 1/b if fc is None else fc
        return c, loc, scale
```

mdhaber commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -3089,6 +3103,32 @@ def test_moments_stats(self, r):
+        data = stats.loglaplace.rvs(c, loc=loc, scale=scale, size=100,
+                                    random_state=rng)
+
+        kwds = {'method': 'mle', 'floc': loc}
+        if fix_c:
+            kwds['fc'] = c
+        if fix_scale:
+            kwds['fscale'] = scale
+        nfree = 4 - len(kwds)
+
+        if nfree == 0:
+            error_msg = "All parameters fixed. There is nothing to optimize."
+            with pytest.raises((RuntimeError, ValueError), match=error_msg):
+                stats.loglaplace.fit(data, **kwds)
+            return
+
+        _assert_less_or_close_loglike(stats.loglaplace, data, **kwds)
```
This part was a little more clear before:
```suggestion
        kwds = {'floc': loc}
        if fix_c:
            kwds['fc'] = c
        if fix_scale:
            kwds['fscale'] = scale
        nfree = 3 - len(kwds)

        if nfree == 0:
            error_msg = "All parameters fixed. There is nothing to optimize."
            with pytest.raises((RuntimeError, ValueError), match=error_msg):
                stats.loglaplace.fit(data, method='mle', **kwds)
            return

        _assert_less_or_close_loglike(stats.loglaplace, data, method='mle', **kwds)
```
Do we even need to pass `method='mle'`, since it's the default?

fancidev replied:
Thanks for your suggestions! I made an update with those.

I put `method='mle'` in case we change the default method in the future. Let's say it's forward looking :) But since 99% of the existing code assumes MLE, it's better to follow them indeed.

mdhaber reviewed (approved):
I think this is all that's left to consider, and I'm comfortable merging as-is if you prefer.

fancidev committed:
MAINT: small code readability improvements.

mdhaber commented on scipy/stats/_continuous_distns.py:
```diff
@@ -6384,6 +6384,37 @@ def _munp(self, n, c):
     def _entropy(self, c):
         return np.log(2.0/c) + 1.0
 
+    @_call_super_mom
```
```suggestion
    @_call_super_mom
    @inherit_docstring_from(rv_continuous)
```

mdhaber committed:
Update scipy/stats/_continuous_distns.py

[skip actions]

mdhaber reviewed (approved):
This is fine but there was a doctest failure because when `fit` is overridden, the docstring needs to be replaced. I'll merge when CI passes.

fancidev commented:
> This is fine but there was a doctest failure because when `fit` is overridden, the docstring needs to be replaced. I'll merge when CI passes.

Ah thanks! I was wondering why all those fit methods have overridden doc string but I’m unable to find the overridden content anywhere in the doc!

mdhaber commented:
>  I’m unable to find the overridden content anywhere in the doc!

You can get them with `help`, but they are not rendered (an problem that is documented in issues).
